File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
n, m = map(int, input().split())
grids = [list(map(int, input().split())) for _ in range(n)]
house = []
chick = []
res = int(1e9)
# 집과 치킨집 좌표들 찾기 (모든 좌표는 1부터 시작함)
for i in range(n):
    for j in range(n):
        if grids[i][j] == 1:
            house.append([i, j])
        elif grids[i][j] == 2:
            chick.append([i, j])

# 조합 구현
def comb(arr, r):
    for i in range(len(arr)):
        if r==1:
            yield [arr[i]]
        else:
            for next in comb(arr[i+1:], r-1):
                yield [arr[i]]+next

# 치킨 집에서 m개 뽑아서 치킨 거리 구하기
for c in comb(chick, m):
    cd = 0  # 현재 치킨 거리(chicken distance)
    for h in house:
        temp = int(1e9)
        for dot in c:   # 치킨 거리 저장(치킨 집과의 가장 짧은 거리)
            temp = min(temp, abs(h[0]-dot[0]) + abs(h[1]-dot[1]))
        cd += temp
    res = min(res, cd)

print(res)

File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
	static int n, m, ans = Integer.MAX_VALUE;
	static int[][] map;
	static ArrayList<Pair> chicken, house;
	static ArrayList<ArrayList<Pair>> candidates;
	static boolean[] visited;
	
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        
        map = new int[n][n];
        for (int i=0; i<n; i++) {
        	st = new StringTokenizer(br.readLine());
        	for (int j=0; j<n; j++) {
        		map[i][j] = Integer.parseInt(st.nextToken());
        	}
        }
        
        // 전체 치킨집, 집 배열에 추가
        chicken = new ArrayList<>();
        house = new ArrayList<>();
        for (int i=0; i<n; i++) {
        	for (int j=0; j<n; j++) {
        		if (map[i][j] == 1) {
        			house.add(new Pair(i, j));
        		}
        		if (map[i][j] == 2) {
        			chicken.add(new Pair(i, j));
        		}
        	}
        }
   
        
        // 가능한 치킨 집 뽑기
        candidates = new ArrayList<>();
        visited = new boolean[chicken.size()];
        combination(0, m);
        
        int sum = Integer.MAX_VALUE;
        for(ArrayList<Pair> candidate : candidates) {
        	int chickenDistance = 0;
            for (Pair h : house) {
            	int temp = Integer.MAX_VALUE;
            	for (Pair c : candidate) {
            		int dm = Math.abs(h.x-c.x) + Math.abs(h.y-c.y);
            		
            		temp = Math.min(temp, dm);
            	}
            	chickenDistance += temp;
            }
            
            sum = Math.min(chickenDistance, sum);
        }
        
        System.out.print(sum);
    }
    
    private static void combination(int n, int r) {
    	if(r == 0) {
            ArrayList<Pair> candidate = new ArrayList<>();
            for(int i = 0; i < chicken.size(); i++) {
                if(visited[i]) {
                    candidate.add(chicken.get(i));
                }
            }
            candidates.add(candidate);
            return;
        } 

        for(int i=n; i<chicken.size(); i++) {
            visited[i] = true;
            combination(i + 1, r - 1);
            visited[i] = false;
        }
    }
    
    static class Pair {
    	int x, y;
    	public Pair(int x, int y) {
    		this.x = x;
    		this.y = y;
    	}
    }
}